File: scripts/optimizer/beasts.py
"""Companions Data -> beasts, their skills and the affection stat table.

The beast list (Tier / Full Name / Skill Description / Skill X / awaken
columns None, 0..6 / 1st and 2nd Effect Name) sits beside a stat table: one
row per affection level, and for each tier (Common, Unique) and awaken level
(0..6) the columns Combat, Draco Combat, ATK/Affection and MSPD, in percent.
"""

from optimizer.workbook import MissingHeader, find_cell, header_columns, number, text

STATS = ["combat", "dracoCombat", "atkAffection", "mspd"]
EFFECTS = {"Increased Attack :": "atk", "Increased MSPD :": "mspd", "Increased Affection :": "affection"}
FAMILIES = ["Wolf", "Boar", "Bat", "Golem", "Draco"]
AWAKENINGS = 7


def _family(name):
    for family in FAMILIES:
        if name.endswith(family):
            return family
    raise ValueError(f"beast {name!r} is not a wolf, boar, bat, golem or draco")


def extract_beasts(sheet):
    """Return {"beasts": [...], "tables": {tier: {stat: [[level 1.. values] per awaken]}}}."""
    header_row, name_col = find_cell(sheet, "Full Name")
    cols = header_columns(sheet, header_row, ["Tier", "Skill Description", "Skill X", "1st Effect Name", "2nd Effect Name"])
    _, none_col = find_cell(sheet, "None", min_row=header_row, max_row=header_row, min_col=cols["Skill X"])

    beasts = []
    row = header_row + 2
    while text(sheet.cell(row, name_col).value):
        name = text(sheet.cell(row, name_col).value)
        effects = [EFFECTS[text(sheet.cell(row, cols[c]).value)] for c in ("1st Effect Name", "2nd Effect Name") if text(sheet.cell(row, cols[c]).value)]
        if not effects:
            raise ValueError(f"{sheet.title} row {row}: {name} has no mounted effect")
        x = sheet.cell(row, cols["Skill X"]).value
        beasts.append({
            "name": name,
            "family": _family(name),
            "tier": text(sheet.cell(row, cols["Tier"]).value),
            "mounted": effects,
            "skill": {
                "text": text(sheet.cell(row, cols["Skill Description"]).value),
                "x": number(x) if number(x) is not None else text(x),
                "values": [number(sheet.cell(row, none_col + 1 + a).value) or 0 for a in range(AWAKENINGS)],
            },
        })
        row += 1
    if not beasts:
        raise MissingHeader(f"{sheet.title}: no beasts under Full Name")

    stat_row, first_col = find_cell(sheet, "Combat")
    level_col = first_col - 1
    tier_row = stat_row - 4
    tables = {}
    for tier in sorted({b["tier"] for b in beasts}):
        _, tier_col = find_cell(sheet, tier, min_row=tier_row, max_row=tier_row, min_col=first_col)
        table = {stat: [] for stat in STATS}
        for awaken in range(AWAKENINGS):
            for offset, stat in enumerate(STATS):
                col = tier_col + awaken * len(STATS) + offset
                values = []
                r = stat_row + 1
                while number(sheet.cell(r, level_col).value) is not None:
                    value = number(sheet.cell(r, col).value)
                    if value is None:
                        break
                    if number(sheet.cell(r, level_col).value) != len(values) + 1:
                        raise ValueError(f"{sheet.title} row {r}: affection levels out of order")
                    values.append(value)
                    r += 1
                table[stat].append(values)
        tables[tier] = table
    return {"beasts": beasts, "tables": tables}

File: src/lib/game/beasts.ts
/** A beast from Companions Data: mounted effects and its skill (Y by awaken level 0..6). */
export type Beast = {
  name: string;
  family: "Wolf" | "Boar" | "Bat" | "Golem" | "Draco";
  tier: string;
  mounted: ("atk" | "mspd" | "affection")[];
  skill: { text: string; x: number | string; values: number[] };
};
type BeastStat = "combat" | "dracoCombat" | "atkAffection" | "mspd";
/** Percent values by tier, stat and awaken level, one per affection level from 1. */
export type BeastData = { beasts: Beast[]; tables: Record<string, Record<BeastStat, number[][]>> };
/** A beast's awaken level (null: not owned) and affection level. */
export type BeastState = { awaken: number | null; affection: number };

export const MAX_BEAST_AWAKEN = 6;

/** Affection is capped at 10 levels per awaken step: 10 at awaken 0, 70 at awaken 6. */
export const maxAffection = (awaken: number) => (awaken + 1) * 10;

/** A table value in percent for an owned beast; 0 when not owned. */
export function beastValue(data: BeastData, beast: Beast, state: BeastState | undefined, stat: BeastStat): number {
  if (!state || state.awaken === null) return 0;
  const levels = data.tables[beast.tier]?.[stat]?.[state.awaken] ?? [];
  const affection = Math.min(Math.max(1, state.affection), levels.length);
  return levels[affection - 1] ?? 0;
}

export type BeastTotals = {
  /** Owned effect on ATK, HP and HP Recovery, as a fraction (Companions Data GM2). */
  combat: number;
  /** Mounted effects as fractions: ATK (GM3, counted only while mounted), MSPD and affection. */
  mountedAtk: number;
  mspd: number;
  affection: number;
};

/**
 * Totals over every owned beast. As in the workbook, the mounted ATK of every
 * attack beast counts once any beast is mounted.
 */
export function beastTotals(data: BeastData, states: Record<string, BeastState>, mounted: boolean): BeastTotals {
  const totals: BeastTotals = { combat: 0, mountedAtk: 0, mspd: 0, affection: 0 };
  for (const beast of data.beasts) {
    const state = states[beast.name];
    totals.combat += beastValue(data, beast, state, beast.family === "Draco" ? "dracoCombat" : "combat") / 100;
    for (const effect of beast.mounted) {
      const value = beastValue(data, beast, state, effect === "mspd" ? "mspd" : "atkAffection") / 100;
      if (effect === "atk") totals.mountedAtk += mounted ? value : 0;
      else totals[effect] += value;
    }
  }
  return totals;
}

/** The skill's text with X and Y filled in for an awaken level. */
export function beastSkillText(beast: Beast, awaken: number | null): string {
  const y = awaken === null ? null : beast.skill.values[awaken];
  return beast.skill.text.replace("X", String(beast.skill.x)).replace("Y", y === null || y === undefined ? "Y" : String(y));
}
